### PR TITLE
Resolve several issues in generated Go code

### LINKED ISSUE #8489: [BUG] [Go] `oneOf` generates uncompilable code for primitive types
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

The go generator creates uncompilable Go code when using oneOf with base JSON types (string, number, ...)

##### openapi-generator version

5.0.0 and current 5.0.1-SNAPSHOT

##### OpenAPI declaration file content or url

Subpath is `#/components/schemas/`

```json
      "uiNodeInputAttributes": {
        "description": "InputAttributes represents the attributes of an input node",
        "properties": {
          "disabled": {
            "description": "Sets the input's disabled field to true or false.",
            "type": "boolean"
          },
          "label": {
            "$ref": "#/components/schemas/uiText"
          },
          "name": {
            "description": "The input's element name.",
            "type": "string"
          },
          "pattern": {
            "description": "The input's pattern.",
            "type": "string"
          },
          "required": {
            "description": "Mark this input field as required.",
            "type": "boolean"
          },
          "type": {
            "$ref": "#/components/schemas/uiNodeInputAttributeType"
          },
          "value": {
            "description": "The input's value.",
            "nullable": true,
            "oneOf": [
              {
                "type": "string"
              },
              {
                "type": "number"
              },
              {
                "type": "boolean"
              }
            ]
          }
        },
        "required": [
          "name",
          "type",
          "disabled"
        ],
        "type": "object"
      },
```
(for JSON code), so it becomes more readable. If it is longer than about ten lines,
please create a Gist (https://gist.github.com) or upload it somewhere else and
link it here.
  -->

##### Generation Details

The generated code is:

```
// UiNodeInputAttributes InputAttributes represents the attributes of an input node
type UiNodeInputAttributes struct {
	// Sets the input's disabled field to true or false.
	Disabled bool `json:"disabled"`
	Label *UiText `json:"label,omitempty"`
	// The input's element name.
	Name string `json:"name"`
	// The input's pattern.
	Pattern *string `json:"pattern,omitempty"`
	// Mark this input field as required.
	Required *bool `json:"required,omitempty"`
	Type string `json:"type"`
	// The input's value.
	Value NullableOneOfstringnumberboolean `json:"value,omitempty"` // This type is unknown and does not compile
}
```

Similar problems occur if nullable is false or if we use refs with the same types instead. Here is another example:

```go
	Value *OneOfstringnumberboolean `json:"value,omitempty"`
```

##### Steps to reproduce

Happy to provide a reproducible test case if required. Code is generated as such:

```
npm run openapi-generator-cli -- generate -i ".schema/api.openapi.json" \
				-g go \
				-o "internal/httpclient" \
				--git-user-id ory \
				--git-repo-id kratos-client-go \
				--git-host github.com \
				-c .schema/openapi/gen.go.yml
```

`gen.go.yml` config is:

```
disallowAdditionalPropertiesIfNotPresent: true
packageName: kratos
generateInterfaces: false
isGoSubmodule: false
structPrefix: true
```

##### Related issues/PRs

Couldn't find any.

##### Suggest a fix

If oneOf can't be properly determined, just use `interface{}`.

### PR DESCRIPTION
**Pull request by aeneasr**
This PR resolves several generated Go-client code issues which resolve compilation errors, improve return and field types (pointer vs value) to better match idiomatic Go-style.

Closes #8489
See #8489
See #8485

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@antihax @grokify @kemokemo @jirikuncar

### CODE REVIEW

**Comment by aeneasr** [comment]
I don't understand why this fails. The go tests run locally fine. Any ideas?

**Comment by aeneasr** [comment]
@antihax @grokify @kemokemo @jirikuncar this patch includes some major improvements to the Go generated client code. It resolves several bugs and improves readability of the consumed go code by using best practice, idiomatic, Go patterns. I would appreciate a review and potentially some help around resolving the CI issue as I can not reproduce it locally!

File: samples/client/petstore/go/go-petstore/README.md
Comment (jirikuncar) [inline samples/client/petstore/go/go-petstore/README.md:25]: it would be nice to configure the sample client to generate correct import name:
```suggestion
import petstore "github.com/OpenAPITools/openapi-generator/samples/client/petstore/go/go-petstore/petstore"
```

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (jirikuncar) [inline modules/openapi-generator/src/main/resources/go/api.mustache:332]: This causes tests to fail since you are returning a reference. We should review this change carefully as it might be backward incompatible.

Anyway, I think we should stick with this patch even if it might cause problems for existing clients.

File: samples/client/petstore/go/go-petstore/README.md
Comment (aeneasr) [reply samples/client/petstore/go/go-petstore/README.md:25]: I would assume that GIT_USER_ID and GIT_REPO_ID are properly set in the generator - as those are the vars being used by the template. Is that not so? For generated clients that do set these values, this works as you suggested.

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/api.mustache:332]: Ah, thanks for the hint, I'll look into this! For some reason running the tests locally returns 403 errors. I have to look into the setup why that is. I assume it's missing a stub server?

And yes, this would be a breaking change. However, the proposed fix follows best practices on Go error handling which should generally be a nil-able pointer, and not a value.

There are also some other smaller breaking changes to the value and pointer handling, and some of them resolve Go compilation issues, so I think it's sensible to merge this even if it has changes around these APIs.

This patch also removes pointer slices (see https://github.com/OpenAPITools/openapi-generator/pull/8491/files#diff-f1bc5f7c50de2d30d0b17d2e43864e9b919efc2785e9bcfeac298c1edd310caeR150) which do not make sense, as slices are already pointers to the underlying memory structure - so we're using a pointer to a pointer here which makes dealing with the generated Go code difficult. Similar things apply for maps as well!

File: samples/client/petstore/go/go-petstore/README.md
Comment (wing328) [reply samples/client/petstore/go/go-petstore/README.md:25]: yes, these are set via options (global).

Comment (aeneasr) [reply samples/client/petstore/go/go-petstore/README.md:25]: Ok, so this is correct then as it is right now?

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/api.mustache:332]: @wing328 could you help out here why I am seeing the 403 errors locally?

**Comment by aeneasr** [comment]
I know that reviewing large PRs is very straining. Unfortunately, the Go generator is simply broken at the moment and this PR addresses so many issues in the generated code. I would be really happy to get some pointers what I need to do next in order to get this PR through.

**Comment by pmcgleenon** [comment]
I'm also getting 405/404 errors when running the integration tests.  When I checked it out, the tests were running against the server "petstore.swagger.io".    

Looking at what happens during the CI script, it adds this entry to /etc/hosts
127.0.0.1       petstore.swagger.io 

Then it starts a docker container "swaggerapi/petstore" which I guess is the local server that we need, so the tests are communicating with the server inside the docker image

Here's the CI script
https://github.com/cognitedata/openapi-generator/blob/fix/cogniteTypes/circle.yml

@aeneasr is it worth a try with a similar setup to see if something obvious is going causing the test failure?

**Comment by wing328** [comment]
@aeneasr Thanks again for the PR. Let me review shortly. We were busy with the v5.1.0 release.

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/api.mustache:7]: What if one of the parameter names in the operation is `context`? Will that result in issues after this change?

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (wing328) [reply modules/openapi-generator/src/main/resources/go/api.mustache:332]: To setup the test server locally, please run 
```
docker pull swaggerapi/petstore
docker run -d -e SWAGGER_HOST=http://petstore.swagger.io -e SWAGGER_BASE_PATH=/v2 -p 80:8080 swaggerapi/petstore
```

and add the following entry to your host table:
```
127.0.0.1    petstore.swagger.io
```

File: modules/openapi-generator/src/main/resources/go/client.mustache
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/client.mustache:450]: Not sure what you mean by this. Any issues using `bytes`?

File: modules/openapi-generator/src/main/resources/go/model_oneof.mustache
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/model_oneof.mustache:4]: Do you have a mininal spec to show the issue related to this fix?

We may better fix it in the Go client generator (java class) instead.

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (pmcgleenon) [reply modules/openapi-generator/src/main/resources/go/api.mustache:332]: @wing328  I've added those steps to the wiki

https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests

File: modules/openapi-generator/src/main/resources/go/client.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/client.mustache:450]: Right, this is a bit confusing. The problem is that in another file `bytes` might or might not be needed as an import. By defining the function here, as bytes is always required then, we avoid unused import / undefined import. This is the same as `fmt.Errorf` but the comment can probably be improved:

```suggestion
// A wrapper for strict JSON decoding
```

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/api.mustache:7]: iirc all methods used are upper-case as they would not be exported otherwise, meaning that `context` can not conflict with one of those. However, I might have missed something - do you have a specific place in mind that could conflict here?

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/api.mustache:332]: Oh great, thank you! I will address this tomorrow!

File: modules/openapi-generator/src/main/resources/go/model_oneof.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/model_oneof.mustache:4]: Yes, so the problem is if we have simple types in the oneOf (bool, float, string) as we generate

```
type UiNodeInputAttributesValue struct {
	bool    *bool
	float32 *float32
	string  *string
}
```

from

```
      "uiNodeInputAttributesValue": {
        "oneOf": [
          {
            "type": "string"
          },
          {
            "type": "number"
          },
          {
            "type": "boolean"
          }
        ]
      },
```

Using the upper case variant here fixes that:

```
type UiNodeInputAttributesValue struct {
	Bool    *bool
	Float32 *float32
	String  *string
}
```

File: modules/openapi-generator/src/main/resources/go/api.mustache
Comment (aeneasr) [reply modules/openapi-generator/src/main/resources/go/api.mustache:7]: Ok, so I looked into this. The thing is, `context` is only used in a few places - primarily in structs, variable declarations, and function signatures:

```
type AdminApiApiCreateIdentityRequest struct {
	ctx            context.Context
	ApiService     *AdminApiService
	createIdentity *CreateIdentity
}
var (
	_ context.Context
)

func (a *AdminApiService) CreateIdentity(ctx context.Context) AdminApiApiCreateIdentityRequest {
	return AdminApiApiCreateIdentityRequest{
		ApiService: a,
		ctx:        ctx,
	}
}
```

In all of those cases there can never be an overshadowing variable clash. So yes, this is safe!

**Comment by aeneasr** [comment]
Thank you for your help! I've addressed everything, and tests now pass also. From my side this is good to go :)

We will be switching to this generator for all of https://github.com/ory/ / https://www.ory.sh and given that we have large and complex api specs I'll very likely come back with more fixes and improvements :)

**Comment by aeneasr** [comment]
I resolved the merge conflicts so this should be good to merge again!

**Review by jirikuncar — APPROVED** [review]

**Comment by jirikuncar** [comment]
Thank you @aeneasr! These are very good improvements :shipit:

**Comment by aeneasr** [comment]
Appreciated :)

**Comment by aeneasr** [comment]
Any chance to get this merged at some point?

**Comment by aeneasr** [comment]
I was hoping this would land in 5.2.0 :/ We've been using this generator for quite some time now and haven't had any issues (e.g. compile issues) so far. There's maybe one or two things that could be improved in pointerland but otherwise this template is working really well!

**Comment by rmangi** [comment]
I just came across this looking for work happening on the go generator. Is it going to get merged in? It's unfortunate to see work sitting and waiting.

**Comment by aeneasr** [comment]
@wing328 happy to resolve the merge conflicts again but it would be nice to see a commitment if this would make it into 5.3.x? :)

**Comment by aeneasr** [comment]
Looks like this just won't get addressed @wing328 , so I'll just go ahead and close this!

**Comment by tbalthazar** [comment]
This PR is a lot of work and has lots of nice improvements, thanks @aeneasr. 
I'm grateful for all the work that has already been put on the project by the maintainers and contributors and I'd love to understand why this hasn't been merged. Has support for the Go generator been discontinued? Or is everybody just super busy? It would help clarify the current status of the Go generator. Cc: @wing328 @jirikuncar

**Comment by aeneasr** [comment]
Thank you @tbalthazar ! I have been asked by several people to reopen the PR in hopes it gets merged at some point. So reopening!

@wing328 I know how difficult it is to maintain such a large project. If there's anything I can do to help and get things resolved please let me know. I'll do my best :)

**Comment by wing328** [comment]
@aeneasr first of all. Thanks again for the PR. I left some feedback before but didn't have time to review your replies.

Can you please merge the latest master into your branch and I'll try to take another look this week?  Thanks again for your work.

**Comment by aeneasr** [comment]
No problem, I'll get on it immediately!

**Comment by sabrina981** [comment]
Do not open my account information up again! Get out loser with candy crush on your own home account code for payment ASAP thanks again for tracking your code bill for tracking payment code

**Comment by aeneasr** [comment]
@wing328 issues resolved, CI passes! It's good to go! :)

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache
Comment (jirikuncar) [inline modules/openapi-generator/src/main/resources/go/model_simple.mustache:20]:
```suggestion
	// Deprecated
```

File: samples/openapi3/client/petstore/go/go-petstore/model_pet.go
Comment (jirikuncar) [inline samples/openapi3/client/petstore/go/go-petstore/model_pet.go:25]:
```suggestion
	// Deprecated
```

**Review by jirikuncar — CHANGES_REQUESTED** [review]
nit: can you please replace spaces with a tab?

**Comment by aeneasr** [comment]
Done :)

**Review by jirikuncar — APPROVED** [review]

**Comment by aeneasr** [comment]
CI passed again, everything green, good to go! :)

**Comment by wing328** [comment]
@jirikuncar thanks for reviewing the change.

@aeneasr thanks again for your work. Can you please PM me via Slack for a few quick questions when you've time? https://join.slack.com/t/openapi-generator/shared_invite/enQtNzAyNDMyOTU0OTE1LTY5ZDBiNDI5NzI5ZjQ1Y2E5OWVjMjZkYzY1ZGM2MWQ4YWFjMzcyNDY5MGI4NjQxNDBiMTlmZTc5NjY2ZTQ5MGM

My timezone is +0800 (Hong Kong).

**Comment by wing328** [comment]
Tested locally and the result is good. Thanks for the PR and sorry again for the delay in reviewing the change

**Comment by aeneasr** [comment]
@wing328 thank you so much for reviewing and merging this and reaching out on Slack!!

**Pull request merged**
